lib/eth/EthRpc.js

```diff
@@ -177,56 +178,184 @@ class EthRPC {
+    const blocks = await this._getRecentBlocks({ cache });
+    const percentileFees = this._getPercentileFees({ percentile, blocks, cache, getFees });
+    if (!percentileFees.length) {
+      return rpcEstimate;
+    }
+    const estimate = this._calculateFeeEstimate(percentileFees);
+    return Math.max(estimate, rpcEstimate);
   }
 
+  /**
+   * Estimates the gas price.
+   * @returns {Promise<number>} The estimated gas price.
+   */
   async estimateGasPrice() {
+    const rpcEstimate = parseInt(await this.web3.eth.getGasPrice());
+    return this._estimateFee({
+      cache: this.blockGasPriceCache,
+      rpcEstimate,
+      percentile: 25,
+    });
+  }
+
+  /**
+   * Estimates the maximum base fee. Defaults to 2 * baseFee + 2.5 gwei.
+   * @param {Object} params - The parameters for estimating the maximum base fee.
+   * @param {number} [params.percentile] - Optional: The maxPriorityFee percentile from last block to use for the estimation.
+   * @param {number} [params.priority] - Optional: The maxPriorityFee to be used with the baseFee.
+   * @returns {number} The estimated maximum base fee.
+   */
+  async estimateMaxFee({ percentile,  priority }) {
+    const lastBlock = await this.web3.eth.getBlock('latest');    
+    const baseFeePerGas = parseInt(lastBlock.baseFeePerGas) ? parseInt(lastBlock.baseFeePerGas) : 0;
+
+    if (priority && parseInt(priority) > 0) {
+      // user defined priority fee
+      return baseFeePerGas + priority;
+    }
+    if (percentile && parseInt(percentile) > 0 && parseInt(percentile) <= 100) {
+      // percentile bases priority fee
+      const maxPriorityFee = await this.estimateMaxPriorityFee({percentile});
+      return baseFeePerGas + maxPriorityFee;
+    }
+    // default max fee formmula. ensures inclusion in next block
+    return 2 * baseFeePerGas + parseInt(this.web3.utils.toWei('2.5', 'gwei'));
+  }
+
+  /**
+   * Estimates the maximum priority fee.
+   * @param {Object} params - The parameters for estimating the maximum priority fee.
+   * @param {number} [params.percentile=25] - The percentile to use for the estimation.
+   * @returns {Promise<number>} The estimated maximum priority fee.
+   */
+  async estimateMaxPriorityFee({ percentile = 25 }) {
+    const rpcEstimate = parseInt(this.web3.utils.toWei('2.5', 'gwei'));
+    const getFees = (_txs) => _txs
+      .filter((tx) => !!tx.maxPriorityFeePerGas)
+      .map((tx) => parseInt(tx.maxPriorityFeePerGas));
+    return this._estimateFee({
+      cache: this.blockMaxPriorityFeeCache,
+      rpcEstimate,
+      percentile,
+      getFees,
+    });
+  }
+
+  /**
+   * Fetches the last n blocks from the blockchain.
+   * @param {Object} params - The parameters for fetching the blocks.
+   * @param {Map} params.cache - The cache to use for storing the blocks.
+   * @param {number} [params.n=10] - The number of blocks to fetch.
+   * @returns {Promise<Array>} The fetched blocks.
+   */
+  async _getRecentBlocks({ cache, n = 10 }) {
```

Comment:
I know this is not your fault but I am wondering if now is a good time to make this method more predictable. It doesn't make sense for it to return block data and fee data it should only return one or the other.

We could pass in a block cache and compare that before making the call to the node, returning only block data. It doesn't make sense for us to have to watch out for block data or fee data from this method.